web3.js/src/connection.ts

```diff
@@ -4117,122 +4161,22 @@ export class Connection {
-            err.message,
-          );
-        }
-      }
-    }
-  }
-
-  /**
-   * @internal
-   */
-  async _unsubscribe(
-    sub: {subscriptionId: SubscriptionId | null},
-    rpcMethod: string,
-  ) {
-    const subscriptionId = sub.subscriptionId;
-    if (subscriptionId != null && typeof subscriptionId != 'string') {
-      const unsubscribeId: number = subscriptionId;
-      try {
-        await this._rpcWebSocket.call(rpcMethod, [unsubscribeId]);
-      } catch (err) {
-        if (err instanceof Error) {
-          console.error(`${rpcMethod} error:`, err.message);
-        }
-      }
-    }
-  }
-
-  /**
-   * @internal
-   */
-  _resetSubscriptions() {
-    Object.values(this._accountChangeSubscriptions).forEach(
-      s => (s.subscriptionId = null),
-    );
-    Object.values(this._logsSubscriptions).forEach(
-      s => (s.subscriptionId = null),
-    );
-    Object.values(this._programAccountChangeSubscriptions).forEach(
-      s => (s.subscriptionId = null),
-    );
-    Object.values(this._rootSubscriptions).forEach(
-      s => (s.subscriptionId = null),
-    );
-    Object.values(this._signatureSubscriptions).forEach(
-      s => (s.subscriptionId = null),
-    );
-    Object.values(this._slotSubscriptions).forEach(
-      s => (s.subscriptionId = null),
-    );
-    Object.values(this._slotUpdateSubscriptions).forEach(
-      s => (s.subscriptionId = null),
-    );
+    this._subscriptionCallbacksByServerSubscriptionId = {};
+    Object.entries(
+      this._subscriptionsByHash as Record<SubscriptionConfigHash, Subscription>,
+    ).forEach(([hash, subscription]) => {
+      this._subscriptionsByHash[hash] = {
+        ...subscription,
+        state: 'pending',
+      };
+    });
   }
 
   /**
    * @internal
    */
-  _updateSubscriptions() {
-    const accountClientSubscriptionIds = Object.keys(
-      this._accountChangeSubscriptions,
-    ).map(Number);
-    const programClientSubscriptionIds = Object.keys(
-      this._programAccountChangeSubscriptions,
-    ).map(Number);
-    const slotClientSubscriptionIds = Object.keys(this._slotSubscriptions).map(
-      Number,
-    );
-    const slotUpdateClientSubscriptionIds = Object.keys(
-      this._slotUpdateSubscriptions,
-    ).map(Number);
-    const signatureClientSubscriptionIds = Object.keys(
-      this._signatureSubscriptions,
-    ).map(Number);
-    const rootClientSubscriptionIds = Object.keys(this._rootSubscriptions).map(
-      Number,
-    );
-    const logsClientSubscriptionIds = Object.keys(this._logsSubscriptions).map(
-      Number,
-    );
-    if (
-      accountClientSubscriptionIds.length === 0 &&
-      programClientSubscriptionIds.length === 0 &&
-      slotClientSubscriptionIds.length === 0 &&
-      slotUpdateClientSubscriptionIds.length === 0 &&
-      signatureClientSubscriptionIds.length === 0 &&
-      rootClientSubscriptionIds.length === 0 &&
-      logsClientSubscriptionIds.length === 0
-    ) {
+  async _updateSubscriptions() {
```

Comment:
I'm finding this method very difficult to review because of the potential for race conditions in state transitions. I think it would be easier to follow if this method was synchronous and was refactored to synchronously transitions subscriptions to "subscribing", "unsubscribing", and deleted states. Then, the async methods for subscribing and unsubscribing can check if a given subscription still needs to be subscribed/unsubscribed before and after making the rpc request.

Reply:
It's a bit of a bird's nest, I know. Hopefully though, my comments and reduced test cases above have convinced you that this update method does _exactly_ what you've asked for! It causes every state transition, synchronously through eager evaluation, then yields in exactly two places by awaiting `this._rpcWebSocket.call(...)` depending on whether we need to subscribe or unsubscribe.

Reply:
Cool, this makes sense now. I think there's still potential for race conditions but I'm more happy now with how you've structured this. Thanks for pushing back thoughtfully. I also wrote up this gist to clarify to myself the yielding behavior of async code: https://gist.github.com/jstarry/9b86287eec9468696d5dd28aa7a6c50a